contracts/strategies/HStrategy.sol

```diff
@@ -0,0 +1,841 @@
+    /// @param burnAmounts minimum amount of tokens got from draining position
+    /// @param erc20Vault_ ERC20 vault of the strategy
+    /// @param uniV3Vault_ UniswapV3 vault of the strategy
+    /// @param uniV3Nft the nft of the position from position manager
+    /// @param tokens_ the addresses of the tokens managed by the strategy
+    /// @return tokenAmounts actual amount of tokens got from draining position
+    function _drainPosition(
+        uint256[] memory burnAmounts,
+        IIntegrationVault erc20Vault_,
+        IUniV3Vault uniV3Vault_,
+        uint256 uniV3Nft,
+        address[] memory tokens_
+    ) internal returns (uint256[] memory tokenAmounts) {
+        tokenAmounts = uniV3Vault_.liquidityToTokenAmounts(type(uint128).max);
+        tokenAmounts = uniV3Vault_.pull(address(erc20Vault_), tokens_, tokenAmounts, "");
+        _compareAmounts(burnAmounts, tokenAmounts);
+        emit BurnUniV3Position(uniV3Nft);
+    }
+
+    /// @notice swapping tokens
+    /// @param amountIn amount of tokens to be swapped
+    /// @param tokenInIndex the index of token to be swapped (0 or 1)
+    /// @param restrictions the restrictions of the amount of tokens to be transferred
+    /// @param erc20Vault_ ERC20 vault of the strategy
+    /// @param tokens_ the addresses of the tokens managed by the strategy
+    /// @return amountsOut actual amount of tokens got from swap
+    function _swapTokensOnERC20Vault(
+        uint256 amountIn,
+        uint256 tokenInIndex,
+        RebalanceTokenAmounts memory restrictions,
+        IIntegrationVault erc20Vault_,
+        address[] memory tokens_
+    ) internal returns (uint256[] memory amountsOut) {
+        {
+            (uint256[] memory tvl, ) = erc20Vault_.tvl();
+            if (tvl[tokenInIndex] < amountIn) {
+                amountIn = tvl[tokenInIndex];
+            }
+        }
+        ISwapRouter.ExactInputSingleParams memory swapParams = ISwapRouter.ExactInputSingleParams({
+            tokenIn: tokens_[tokenInIndex],
+            tokenOut: tokens_[tokenInIndex ^ 1],
+            fee: pool.fee(),
+            recipient: address(erc20Vault_),
+            deadline: restrictions.deadline,
+            amountIn: amountIn,
+            amountOutMinimum: 0,
+            sqrtPriceLimitX96: 0
+        });
+
+        bytes memory routerResult;
+        if (amountIn > 0) {
+            bytes memory data = abi.encode(swapParams);
+            erc20Vault_.externalCall(tokens_[tokenInIndex], APPROVE_SELECTOR, abi.encode(address(router), amountIn)); // approve
+            routerResult = erc20Vault_.externalCall(address(router), EXACT_INPUT_SINGLE_SELECTOR, data); // swap
+            erc20Vault_.externalCall(tokens_[tokenInIndex], APPROVE_SELECTOR, abi.encode(address(router), 0)); // reset allowance
+            uint256 amountOut = abi.decode(routerResult, (uint256));
+            require(restrictions.swappedAmounts[tokenInIndex ^ 1] <= amountOut, ExceptionsLibrary.LIMIT_UNDERFLOW);
```

Comment:
why not putting this restriction into AmountOutMinimum?

Reply:
We have decided not to put this restriction into the `amountOutMinimun` field to handle the underflow exception on our side.